### PR TITLE
BUG: memory leak in np.dot of size 0

### LINKED ISSUE #9615: Empty array operations result in significant memory leaks
This is likely an edge case, but empty array operations result in very significant memory leaks.

The following code

    for i in range(int(1e7)):
        np.dot(np.array([]), np.array([]))

consumed >1G of memory and cannot be freed. This is tested on macOS 10.12 and numpy is compiled against OpenBLAS.

### PR DESCRIPTION
**Pull request by ExpHP**
Fixes #9615

### CODE REVIEW

**Comment by ExpHP** [comment]
This should have a regression test but I'm still trying to come up with one that's portable.  (Unfortunately, I cannot just check `sys.getrefcount()` on the array returned by `np.dot` since that appears to be a different object from the one leaked; it already had a refcount of 1 prior to the fix...)

**Comment by ExpHP** [comment]
test added as separate commit for easy verification that it did, in fact, used to fail.  It is a bit convoluted, but documented.

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -2497,6 +2497,35 @@ def test_diagonal_memleak(self):
         if HAS_REFCOUNT:
             assert_(sys.getrefcount(a) < 50)
 
+    def test_size_zero_memleak(self):
+        # Regression test for issue 9615
+
+        orig = np.zeros((6*4+1,), dtype='int8', order='c')[1:]
+        orig.dtype = 'f4'
+        orig[:] = 5
+        copy = np.nditer(orig, [], [['readwrite', 'updateifcopy', 'aligned']]).operands[0]
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: Hmm, we have been trying to get rid of updateifcopy. @mattip ?

**Comment by charris** [comment]
The fix is simple enough that I'd be willing to put this in without the test. But it seems that the total refcount, or count of objects should go up after excuting a loop. We do have a travis tests with a debug build of python if that helps. @njsmith Thoughts?

I'd make the test commit comment something like
```
TST: Test for memory leak in np.dot of size 0 arrays.
```
as that stands by itself. If you want to reference the PR, do that in the comment body.

**Comment by ExpHP** [comment]
> But it seems that the total refcount, or count of objects should go up after excuting a loop.

I just realized how this test could be made much, much, much simpler

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (ExpHP) [reply numpy/core/tests/test_multiarray.py]: this will be eliminated shortly

File: numpy/core/tests/test_multiarray.py [hunk 2]
```diff
@@ -2497,6 +2497,15 @@ def test_diagonal_memleak(self):
         if HAS_REFCOUNT:
             assert_(sys.getrefcount(a) < 50)
 
+    def test_size_zero_memleak(self):
+        # Regression test for issue 9615
+        a = np.array([], dtype='float32')
+        x = np.array(2.0, order='c', dtype='float32')
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: C order is the default, and for a scalar array it doesn't matter ;)

File: numpy/core/tests/test_multiarray.py [hunk 3]
```diff
@@ -2497,6 +2497,15 @@ def test_diagonal_memleak(self):
         if HAS_REFCOUNT:
             assert_(sys.getrefcount(a) < 50)
 
+    def test_size_zero_memleak(self):
+        # Regression test for issue 9615
+        a = np.array([])
```
Comment (charris) [inline numpy/core/tests/test_multiarray.py]: I did like the dtype last time, as it makes clear we are testing floats, so a simple
`np.array([], dtype=np.float64)` would be good.

**Comment by charris** [comment]
LGTM except for the one tiny nit.

File: numpy/core/tests/test_multiarray.py (hunk 3 above)
Comment (charris) [reply numpy/core/tests/test_multiarray.py]: Might add a comment explaining the need for float because the problem was in `cblasfuncs.c`.

**Comment by ExpHP** [comment]
Yeah, just putting the dtype there without any indication of its
significance felt odd

**Comment by charris** [comment]
Thanks @ExpHP .

**Pull request merged**